### PR TITLE
Week 4

### PR DESCRIPTION
**Pull request by FeelFried**

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+class SessionsController < ApplicationController
+  def new; end
```
Comment (alg) [inline app/controllers/sessions_controller.rb:2]: Empty methods can be deleted.

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        end
+      else
+        flash.alert = 'Name is invalid'
+         render :new
+      end
+    else
+      flash.alert = 'Email or name is already taken'
+       render :new
+    end
+  end
```
Comment (alg) [inline app/controllers/users_controller.rb:32]: All these validations should go into User model as `validates ...` statements. Ideally, the controller should look like:
```
  def new
    @user = User.new
  end

  def create
    @user = User.new(user_params)
    if @user.save
      flash[:success] = 'Registration Successful!'
      redirect_to root_path
    else
      render :new
    end
  end

  private

  def user_params
    params.require(:user).permit(:name, :email, :password, :password_confirmation)
  end
```

File: app/views/layouts/application.html.erb [hunk 3]
```diff
@@ -1,4 +1,4 @@
-<!DOCTYPE html>
+<!DOCTYPE html>0
```
Comment (alg) [inline app/views/layouts/application.html.erb:1]: What is this "0"?

File: app/views/layouts/application.html.erb [hunk 4]
```diff
@@ -8,9 +8,17 @@
...
   </head>
 
-  <body>
+  <body class="bg-dark">
+    <%= render partial: 'layouts/navbar' %>
+    <% if flash[:notice].present? %>
+      <div class="alert alert-success" role="alert">
+        <%= flash[:notice] %>
+      </div>
+    <% end %>
```
Comment (alg) [inline app/views/layouts/application.html.erb:21]: You should move your `flash[:alert]` here too, so it's not duplicated across all pages.

File: app/views/sessions/_login_form.html.erb [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+				</div>
+			</div>
+			<div class="row my-3">
+				<div class="col-12">
+					<%= submit_tag 'Login', class: 'btn btn-secondary' %>
+				</div>
+			</div>
+
+		<% end %>
+	</div>
```
Comment (alg) [inline app/views/sessions/_login_form.html.erb:36]: We keep our syntax consistent and using 2 spaces for indentation.

File: app/views/users/_signup_form.html.erb [hunk 6]
```diff
@@ -0,0 +1,59 @@
+<br/>
+<br/>
+<br/>
+	<div class="col-12 offset-12 col-sm-4 offset-sm-4  text-white">
+		<% if flash[:alert] %>
+			<div id="alert" class="text-danger" >
+				 <%= flash[:alert] %>
+			</div>
+		<% end %>
+		<%= form_tag signup_url, class: 'form' do %>
```
Comment (alg) [inline app/views/users/_signup_form.html.erb:10]: It's better to create an empty user in the `UsersController#new` action and use it here via `form_for @user`. Then if there are any validation errors, you pass it back to this template in the `@user` instance var and the form is rendered with values prefilled. You also have your errors associated with each user field in `@user.errors['field']` or `@user.errors.full_messages_for('field')` that you can render near the field.

File: app/views/sessions/new.html.erb [hunk 7]
```diff
@@ -0,0 +1 @@
+<%= render partial: 'login_form'  %>
```
Comment (alg) [inline app/views/sessions/new.html.erb:1]: You are not going to reuse this form and the template doesn't look busy, then why extracting a partial? If you paste it here from the `login_form` file, it won't become any more complex.

File: app/views/users/new.html.erb [hunk 8]
```diff
@@ -0,0 +1 @@
+<%= render partial: 'signup_form'  %>
```
Comment (alg) [inline app/views/users/new.html.erb:1]: Same as in `sessions/new.html.erb`. You don't need to extract the form before you need to reuse it.

File: config/routes.rb [hunk 9]
```diff
@@ -1,3 +1,10 @@
 Rails.application.routes.draw do
-  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  get 'users/new'
```
Comment (alg) [inline config/routes.rb:2]: You probably don't need it.

File: db/migrate/20210520162624_create_users.rb [hunk 10]
```diff
@@ -0,0 +1,11 @@
+class CreateUsers < ActiveRecord::Migration[6.1]
+  def change
+    create_table :users do |t|
+      t.string :name, null: false, index: { unique: true }
```
Comment (alg) [inline db/migrate/20210520162624_create_users.rb:4]: Names should be unique? That's ok, just don't see why.

File: test/fixtures/users.yml [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+
+one:
+  name: MyString
+  email: MyString
+  password_digest: MyString
+
+two:
+  name: MyString
+  email: MyString
+  password_digest: MyString
```
Comment (alg) [inline test/fixtures/users.yml:11]: Please use FactoryBot instead of fixtures.

File: test/controllers/users_controller_test.rb [hunk 12]
```diff
@@ -0,0 +1,8 @@
+require "test_helper"
+
+class UsersControllerTest < ActionDispatch::IntegrationTest
+  test "should get new" do
+    get users_new_url
+    assert_response :success
+  end
```
Comment (alg) [inline test/controllers/users_controller_test.rb:7]: I guess it's a practice test.

File: db/migrate/20210520162624_create_users.rb (hunk 10 above)
Comment (FeelFried) [reply db/migrate/20210520162624_create_users.rb:4]: Will be used for login in future ( login by email/user name ), mb better add unique nicknames for that purpose, I haven't decided yet.

File: app/models/user.rb [hunk 13]
```diff
@@ -0,0 +1,8 @@
+class User < ApplicationRecord
+  has_secure_password
+  email_regex = /\A([\w+\-].?)+@[a-z\d\-]+(\.[a-z]+)*\.[a-z]+\z/i
```
Comment (alg) [inline app/models/user.rb:3]: `email_regex` is a constant. Constant names are capitalized in Ruby -- `EMAIL_REGEX`.

You can simplify it slightly:

```
/\A([\w+\-].?)+@[a-z\d\-]+(\.[a-z]+)+\z/i
```

File: app/views/layouts/application.html.erb [hunk 14]
```diff
@@ -5,12 +5,25 @@
...
+        <%= flash[:notice] %>
+      </div>
+      <% end %>
+      <% if flash[:alert].present? %>
+        <div id="alert" class="text-danger">
+          <%= audio_tag "/assets/alert.wav", autoplay: true%>
+          <%= flash[:alert] %>
+        </div>
+      <% end %>
+    </div>
```
Comment (alg) [inline app/views/layouts/application.html.erb:26]: Consider extracting alerts block into partial for clarity.

**Pull request merged**